layers: Add an Android default log force option via setprop

neelbedekar opened the pull request:
It's useful for certain downstream tools/systems(like automation)
to parse the log output of validation layers. On Android via logcat,
the default validation layer callback only prints when the Vulkan app
explicitly doesn't subscribe to the VK_EXT_debug_utils callback via
vkCreateDebugUtilsMessengerEXT. When the app does subscribe, then only
the app's callback is executed. To allow such downstream tools to
parse logcat for validation layer errors, this PR adds a mechanism to
force the default log to always be printed, regardless of whether the
app subscribes via vkCreateDebugUtilsMessengerEXT. This PR is currently
limited to Android-only, and opt-in with an Android system property.
This allows downstream tools to set the property and parse the logcat
in a known form(accessing key info like VUID), and every other existing
app relying on validation layers will remain unaffected.

ncesario-lunarg commented on layers/vk_layer_logging.h:
```diff
@@ -261,6 +264,23 @@ typedef struct _debug_report_data {
 
+#if defined __ANDROID__
+template <typename T>
+static bool GetSystemProperty(const char *keyName, T *val) {
```
Maybe just return an [Optional](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/b1a10a4bc60940e269b3c3d32d567e689e66e707/layers/vk_layer_data.h#L703)`<T>` and avoid the output pointer?

ncesario-lunarg commented on layers/vk_layer_logging.h:
```diff
@@ -463,6 +483,12 @@ static inline bool debug_log_msg(const debug_report_data *debug_data, VkFlags ms
     }
 
+#if defined __ANDROID__
+    if (debug_data->forceDefaultLogCallback) {
```
If this ends up being true, no need to loop through the callback_list above, correct?

neelbedekar replied:
Yep, no need. I just left that loop unchanged because it's an almost zero-cost operation, with the loop typically being ~2 iterations, and I figured it would be tougher to read from a code cleanliness standpoint if I tried avoiding the loop on Android if debug_data->forceDefaultLogCallback was true.

ncesario-lunarg reviewed (commented):
I think this LGTM with just a couple minor nits. I would like to get some input from others, though, as we would need to support this going forward if added.

If this does end up getting merged, some documentation would be much appreciated. Perhaps either in `BUILD.md`, or perhaps a dedicated Android doc under `docs`?

neelbedekar commented:
Agreed re: documentation, it would be great for us to highlight this functionality in the docs, letting others know that they can set a system property on Android to force the validation layer callback to always log.

I'd be happy to put the docs info wherever you think is best.

sfricke-samsung commented on layers/vk_layer_logging.h:
```diff
@@ -261,6 +264,23 @@ typedef struct _debug_report_data {
 
+#if defined __ANDROID__
+template <typename T>
+static bool GetSystemProperty(const char *keyName, T *val) {
```
So there is a `GetEnvironment` in `vk_layer_config.cpp` that does the `getprop` via `popen`

Nothing against using `__system_property_get`, but really just don't want a second "get prop value from Android function" around for code duplication sake

sfricke-samsung replied:
example we use it for is

```cpp
string version_env = GetEnvironment("ro.build.version.sdk");
int target_version = atoi(version_env.c_str());
```

to know if people are using pre-AHardwareBuffer version of Android

neelbedekar replied:
Oh thanks that's great, I'll gladly use the existing version.

sfricke-samsung commented on layers/vk_layer_logging.h:
```diff
@@ -62,7 +62,9 @@
 #include <android/log.h>
+#include <sys/system_properties.h>
 #define LOGCONSOLE(...) ((void)__android_log_print(ANDROID_LOG_INFO, "VALIDATION", __VA_ARGS__))
+static const char DECORATE_UNUSED *kForceDefaultCallbackKey = "debug.vulkan.debuglogforce";
```
is this a used property name else where? If not can we add `validation` or `vvl` or something to make sure if its unique to validation layers forcing logging that the next person coming through will not think this is going to do other stuff (or worst if some other code on the device is using this)

neelbedekar replied:
Agreed, it's not used elsewhere, but vvl would be good to ensure uniqueness. The length limit is 32 characters, so I'll probably settle on "debug.vvl.forcelayerlog" if that's ok with you.

sfricke-samsung replied:
Sounds good, I think `debug.vvl.` is a good namespace to take!

Only other suggestion is to add a small notice of it in
https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/docs/khronos_validation_layer.md#vk_ext_debug_utils

neelbedekar commented:
Thanks for the review @sfricke-samsung. I updated with a new commit, which mainly switched to GetEnvironment(just had to add a declaration in vk_layer_config.h), used our agreed-upon debug.vvl namespace with a property of debuglogforce, and added a few lines of documentation

sfricke-samsung reviewed (approved):
Looks good to me and documentation seems solid, will let @ncesario-lunarg give final :+1:

ncesario-lunarg reviewed (approved):
LGTM, thanks @neelbedekar and @sfricke-samsung for reviewing! Will merge once updated changes finish internal CI.

ncesario-lunarg commented:
Running through [internal CI](http://erusea:4040/8703).
